File: frontend/src/types/transaction.ts
export type TransactionType = "income" | "expense";

export const PAYMENT_METHODS = [
  "Cash",
  "UPI",
  "Credit Card",
  "Debit Card",
  "Bank Transfer",
  "Other",
] as const;

export type PaymentMethod = (typeof PAYMENT_METHODS)[number];

// Matches the backend's TransactionOut schema exactly - no user_id
// (ownership is implicit), and category_id/payment_method are both
// optional since the backend allows null for either.
export interface Transaction {
  id: number;
  category_id: number | null;
  amount: number;
  type: TransactionType;
  description: string;
  date: string; // ISO date string, e.g. "2026-08-11"
  payment_method: PaymentMethod | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export type TransactionInput = Omit<
  Transaction,
  "id" | "created_at" | "updated_at"
>;

File: backend/app/schemas/transaction.py
from datetime import date as date_type
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

TransactionType = Literal["income", "expense"]
PaymentMethod = Literal["Cash", "UPI", "Credit Card", "Debit Card", "Bank Transfer", "Other"]


class TransactionBase(BaseModel):
    amount: float = Field(..., gt=0, description="Amount must be greater than zero")
    type: TransactionType
    category_id: int | None = None
    description: str = Field(..., min_length=1, max_length=255)
    date: date_type
    payment_method: PaymentMethod | None = None
    notes: str | None = Field(default=None, max_length=500)


class TransactionCreate(TransactionBase):
    pass


class TransactionUpdate(TransactionBase):
    pass


class TransactionOut(TransactionBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    created_at: datetime
    updated_at: datetime
